_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+not the only flavor operators come in.
+The last row of the table has items that to some users may not seem operator-like at all.
+`delete` and `new`  are in fact operators, and follow precedence and associativity rules despite not being symbols.
+Member access via `.`, subscripting arrays with `[]`, or making a function call by adding parentheses
+to a function can be thought of as applications of operators as well.
+Even putting something in parentheses creates a valid operator expression.
+
+It is important to note that in many languages assignment is a statement that is handled specially.
+In Solidity, like in the C-like languages it took inspiration from, assignment can be freely used in
+expressions on par with other operators.
+
+In most cases built-in operators are only defined for [value types](https://docs.soliditylang.org/en/latest/types.html#value-types).
+Notable exceptions are the simple assignment and the "miscellaneous" group.
+For example, you can put almost anything in parentheses or branches of the ternary operator, as a part
+of a bigger expression, even if its type cannot be named or even assigned to anything.
+
+#### `using for`
+Since operators are a purely syntactic element, you can always use ordinary functions as a substitute.
+Solidity has a mechanism that is often used this way:
+[`using for`](https://docs.soliditylang.org/en/latest/contracts.html#using-for).
+
+The directive `using ... for ...` can be used to attach functions to a type.
+These functions can then be used as members on any object of that type and receive the object as
+their first parameter.
+
+```solidity
+pragma solidity ^0.8.13;
+
+type Int is int;
+using {add} for Int global;
+
+function add(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function test(Int a, Int b) pure returns (Int) {
+    return a.add(b);
+}
+```
+
+Currently, `using for` has 3 forms:
+1. `using L for T`: attaches all non-private functions from library `L` to type `T`.
+1. `using L for *`: attaches all non-private functions from library `L` to any type, including types that
+    cannot be named in the language and used explicitly (e.g. types of literals or array slices).
+1. `using {f, L.g, ...} for T`: attaches given free or library functions `f`, `L.g`, ... to type `T`.
+
+The directive can be used:
+- ✅ At the file level.
+- ✅ Inside contract and library definitions.
+
+Can not be used:
+- ❌ Inside interface definitions.
+
+When used inside contracts, it is not inherited.
+Its effects extend only to the base contract that has it and not to any other contracts derived from it.
+
+At the file level, the directive affects all the free functions, contracts and libraries defined within that file.
+It does not affect other files that import the file containing it, unless it is marked as `global`.
+`global` extends its effect to all files where the type is visible.
```

Comment:
Also to files where it is invisible. It is usable on all expressions of that type, it does not matter that the type is visible via name.

Reply:
I mean this is the same with contracts already: Functions on a contract can be called on all expressions of that contract type, you do not need the type of the contract visible.

Reply:
Very good point. I did not think of that. Changed to just `all files`.